[McNaBry] iP

McNaBry opened the pull request:
# Bryan, A Duke Chatbot.

> Hi. I'm Bryan, <br> What can I do for you?

Bryan frees your mind of having to remember things you need to do. It's,

- Text-based. _No fancy pants GUI._
- Easy to learn
- ~~FAST~~ SUPER EASY to use :)

All you need to do is:

1. Download it from [here](https://github.com/McNaBry/ip/releases/tag/A-Jar)
2. Double-click the .jar file. (Or for the tech-savvy, open your terminal and run `java -jar duke.jar`
3. Add your tasks.
4. And let it manage your tasks for you.

It even remembers all the tasks you created before! How **awesome** is that.
And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] A nice GUI. (coming soon) - _Don't worry, its still text-based_

#

If you wish to add a new feature to Bryan, let's say re-arranging your task order:

1. Add your new command to the `Parser` class.
2. Create a new `RearrangeTaskCommand` class that extends the `Command` class
3. Start coding!

Here's how the `Command` abstract class looks like:
```java
public abstract class Command {
    public abstract void execute(
            TaskList tasks, Storage storage, Ui ui) throws DukeException;

    public boolean isExit() {
        return false;
    }
}
```
Override the `execute` function and fill in your logic :smile:

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

McNaBry committed:
ignore .class files

McNaBry committed:
Chatbot initial skeleton

McNaBry committed:
chatbot echo function

McNaBry committed:
chatbot storage and listing function

McNaBry committed:
chatbot marks tasks as done/undone

McNaBry committed:
support for todo, event & deadline

McNaBry committed:
added coloring for task type

McNaBry committed:
add testing for level 0-4

McNaBry committed:
error handling for invalid command and argument for mark/unmark

McNaBry committed:
add error handling for invalid task creation

McNaBry committed:
expanded scope of testing to include errors

McNaBry committed:
add test to mark deadline and event

McNaBry committed:
Replace task array with arrayList

McNaBry committed:
Add delete task function

McNaBry committed:
Add string representation for each task when saved to file

McNaBry committed:
Add save tasks feature

Chatbot now writes the list of tasks to a file when a new task is added, deleted or marked/unmarked.

Chatbot also reads from the file when starting and loads the tasks list.

EXPECTED.txt is updated to handle the new changes.

McNaBry committed:
Handle erroneous file format

Checks the length of the parsed result for each line of the file read.

If the format is not correct, the line is skipped and a task will not be created.

McNaBry committed:
Change chatbot parser

Deadline and event now explicitly splits by their respective delimiters (/from, /to, /by)

Change EXPECTED.txt to reflect this change

McNaBry committed:
Create a DateParser class

Create a date parser class to handle parsing dates given by the user and the date stored in the file

DateParser also handles displaying the date and converting the date to a string that can be stored on file

McNaBry committed:
Update testing for date time

Updated the Event class and its relevant methods to use LocalDateTime.

Update EXPECTED.TXT to reflect the new changes in using datetime.

McNaBry committed:
Update invalid date message

McNaBry committed:
Create Storage class

McNaBry committed:
Abstract Parser, TaskList, *Command, Ui

McNaBry committed:
Separate files into packages

McNaBry committed:
Merge branch 'add-gradle-support'

McNaBry committed:
Add gradle checkstyle

McNaBry committed:
Add skeleton for gradle test

McNaBry committed:
JUnit test for Parser, Ui, TaskList

McNaBry committed:
Update error messages

Update invalid date error messages for event and deadline.

Trim and strip multiline error message in DukeException to standardize error printing

Updated EXPECTED.TXT to reflect these changes.

McNaBry committed:
Add find function

McNaBry committed:
Javadocs for all the necessary methods

McNaBry committed:
Updated javadocs for some variables.

McNaBry committed:
Fix errors thrown by javadoc

import necessary dependencies to allow @link to detect the correct class.

McNaBry committed:
Javadoc for DukeException

McNaBry committed:
Format according to code standard

McNaBry committed:
Merge branch 'branch-A-CodingStandard'

McNaBry committed:
Merge branch 'branch-Level-9'

McNaBry committed:
Javadocs for Find Command

McNaBry committed:
Bug fix and tidy javadocs

lululwtv commented on src/main/java/common/DateParser.java:
```diff
@@ -0,0 +1,88 @@
+ * and displaying {@link LocalDateTime} as "MMM dd yyyy h:mma".
+ * Example: Aug 10 2023 6:00PM.
+ */
+public class DateParser {
```
I like your class naming!

lululwtv commented on src/main/java/data/exception/DukeException.java:
```diff
@@ -0,0 +1,40 @@
+ * The custom {@link Exception} class for Duke.
+ * Used by {@link ui.Ui} to print error messages.
+ */
+public class DukeException extends Exception {
```
Can the exception name be more specific and coherent?

McNaBry replied:
I wrote this class to handle all Duke-related exceptions so I don't really know how to make the name more specific. Would something like `DukeErrorException` or `DukeParseException` be better?

lululwtv commented on src/main/java/common/DateParser.java:
```diff
@@ -0,0 +1,88 @@
+     * The {@link DateTimeFormatter} used to parse
+     * date strings into {@link LocalDateTime}.
+     */
+    private static final DateTimeFormatter parseFormat = 
```
I like that you used line break for every element you are appending! Very readable

McNaBry committed:
Enabled checkstyle

StevenLiudw commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+    public abstract void execute(
+            TaskList tasks, Storage storage, Ui ui) throws DukeException;
+
+    public boolean isExit() {
+        return false;
+    }
```
The isExit() method can be further documented to explain its use case. Is it to determine if a command should lead to an exit?

StevenLiudw commented on src/main/java/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,46 @@
+        Task dl = new Deadline(description, deadline);
+        tasks.add(dl);
+        storage.update(tasks);
+        ui.displayMsg(new String[] {
+            "Okie! I've added a new " + Ui.cTxt("DEADLINE", Ui.Color.BLUE) + ":",
+            "  " + dl.toString(),
+            "Total no. of tasks stored: " + tasks.getSize()
+        });
```
In the execute method, the message for the UI might benefit from being moved to the Ui class for better organization and separation of concerns. That way, the command focuses solely on its logic and leaves the display details to the Ui class.

StevenLiudw commented on src/main/java/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public void execute(
+            TaskList tasks, Storage storage, Ui ui) throws DukeException {
+        Task removedTask = tasks.delete(taskIndex - 1);
+        ui.displayMsg(new String[] {
+            "Okie! I've deleted task " + taskIndex + ":",
+            "  " + removedTask.toString(),
+            "Total no. of tasks stored: " + tasks.getSize()
+        });
```
Similar to the DeadlineCommand, consider moving UI-related messages to the Ui class.

StevenLiudw commented on src/main/java/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,48 @@
+    @Override
+    public void execute(
+            TaskList tasks, Storage storage, Ui ui) throws DukeException {
+        Task removedTask = tasks.delete(taskIndex - 1);
```
The execute method should consider handling scenarios where the provided index might be out of bounds.

StevenLiudw commented on src/main/java/common/DateParser.java:
```diff
@@ -0,0 +1,88 @@
+     * @param date The date string to be parsed.
+     * @return A {@link LocalDateTime} instance.
+     */
+    public static LocalDateTime parseDateString(String date) {
+        try {
+            return LocalDateTime.parse(date, parseFormat);
+        } catch (DateTimeParseException e) {
+            return null;
+        }
+    }
+
```
Returning null in the parseDateString method might not be the best approach. Consider throwing a custom exception instead, which can provide more descriptive error messages to the user.

StevenLiudw commented on src/main/java/common/DateParser.java:
```diff
@@ -0,0 +1,88 @@
+            ))
+            .append(DateTimeFormatter.ofPattern(
+                "[d MMM yyyy]" + "[MMM d yyyy]",
+                Locale.ENGLISH
```
Good use of Locale.ENGLISH to ensure consistency in date formats regardless of the system's locale.

StevenLiudw commented on src/main/java/data/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Override
+    public String toString() {
+        String taskListString = "";
+        for (Task task : this.tasks) {
+            taskListString += task.toFileFormatString() + "\n";
+        }
+        return taskListString.strip();
+    }
+}
```
In the toString method, consider using a StringBuilder for concatenation in loops for better performance.

StevenLiudw commented on src/main/java/data/tasks/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+import java.time.LocalDateTime;
+
+import common.DateParser;
+
+public class Deadline extends Task {
+    private LocalDateTime deadline;
+
+    public Deadline(
+            String detail,
+            LocalDateTime deadline) {
+        super(detail);
+        this.deadline = deadline;
+    }
+
+    public Deadline(
+            String detail,
+            LocalDateTime deadline,
+            boolean isDone) {
+        super(detail, isDone);
+        this.deadline = deadline;
+    }
+
+    @Override
+    public String toString() {
+        return String.format(
+            "[D]%s (by: %s)",
+            super.toString(),
+            DateParser.toDisplayString(this.deadline)
+        );
+    }
+
+    @Override
+    public String toFileFormatString() {
+        return String.format(
+            "D|%s|%s",
+            super.toFileFormatString(),
+            DateParser.toFileString(this.deadline)
+        );
+    }
+}
```
Consider adding JavaDoc for the constructors for clarity.

StevenLiudw commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,305 @@
+            });
+        }
+
+        return new MarkCommand(parseArr[1], type == "mark");
```
In the parseMarkCommand() method, comparing strings using == (like type == "mark") is not ideal. Instead, use type.equals("mark") for string comparisons.

StevenLiudw reviewed (commented):
LGTM. Overall the structure is quite organized. Just a few tiny points that need to improve. :)

McNaBry committed:
Initial migration to JavaFX

McNaBry committed:
Fix label height in JavaFX

Issue with label's height truncating and cutting off content. Fixed by removing prefHeight in VBox

McNaBry committed:
Add basic assertions

Lack of assertions.

Assertions help to detect bugs at runtime which are not meant to occur.

Basic assertions are added in multiple classes to ensure the logic implemented produces the correct results.

McNaBry committed:
Improve general code quality

Improve code quality according to standards learnt over previous weeks.

Improvements include:
* Marking variables as final - for verbosity
* Using StringBuilder
* Adding additional defense clauses and sanitizing input

McNaBry committed:
Merge pull request #2 from McNaBry/branch-A-Assertions

Add basic assertions

McNaBry committed:
Merge branch 'master' into branch-A-CodeQuality

McNaBry committed:
remove multiline text block

McNaBry committed:
Add InvalidParamException

No custom exception created for invalid parameters passed in by user.

By creating one, the type of exception is thrown is clearer.

McNaBry committed:
Add UnrecognizedCommandException

No custom exception created for unrecognized command passed in by user.

By creating one, the type of exception is thrown is clearer.

Redundant parameters such as uiCli were also removed.

McNaBry committed:
Add StorageException

No custom exception created for exceptions thrown by the Storage class.

By creating one, the type of exception is thrown is clearer.

Additionally, methods that throw exceptions more specific than DukeException now indicate the new exceptions thrown.

McNaBry committed:
Merge pull request #3 from McNaBry/branch-A-CodeQuality

Improve general code quality

McNaBry committed:
Add sorting function

Duke now automatically sorts the tasks via date

McNaBry committed:
GUI overhaul

McNaBry committed:
improve GUI

- add special color for user
- add special color for duke error
- add proper exit for javafx

McNaBry committed:
Set stage min height and width

McNaBry committed:
Add product screenshot

McNaBry committed:
Adjust bg radius

adjusted the bg radius of user input and send button to match that of dialog.

McNaBry committed:
Add more error handling

- Checks if start date > end date for events

- Checks if invalid date was stored on file. Storage will assume LocalDateTime.now() instead.

- Checks if there is write permission when updating local file and throws a custom error message.

McNaBry committed:
Merge branch 'master' into branch-A-UserGuide

McNaBry committed:
Initial user guide

McNaBry committed:
Minor grammar changes to UG

McNaBry committed:
minor changes for v2.0

McNaBry committed:
Add release link to UG

McNaBry committed:
Fix write permission

Storage did not check whether file exists before checking write permission.

This caused missing data.txt files to be treated as missing write permission.

McNaBry committed:
Add note for features

McNaBry committed:
Merge branch 'branch-A-UserGuide'
